File: src/chapters/iterative-methods/content/code/gauss-seidel.cpp
#include <vector>
#include <cmath>
#include <iostream>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

// Solve A x = b by Gauss-Seidel iteration.
Vec gauss_seidel(const Mat& A, const Vec& b, double tol = 1e-10, int max_iter = 200) {
    int n = b.size();
    Vec x(n, 0.0);
    for (int k = 0; k < max_iter; ++k) {
        double diff = 0.0;
        for (int i = 0; i < n; ++i) {
            double s = b[i];
            for (int j = 0; j < n; ++j)
                if (j != i) s -= A[i][j] * x[j];   // x already holds updated entries
            double xi = s / A[i][i];
            diff = max(diff, fabs(xi - x[i]));
            x[i] = xi;
        }
        if (diff <= tol) break;
    }
    return x;
}

int main() {
    Mat A = {{4, 2, -1}, {5, -10, 2}, {-2, 3, -7}};
    Vec b = {9, 8, 3};
    Vec x = gauss_seidel(A, b);
    cout << "x = ";
    for (double v : x) cout << v << " ";
    cout << "\n";
}

File: src/chapters/iterative-methods/content/code/gauss-seidel.rs
// Solve A x = b by Gauss-Seidel iteration.
fn gauss_seidel(a: &[Vec<f64>], b: &[f64], tol: f64, max_iter: usize) -> (Vec<f64>, usize) {
    let n = b.len();
    let mut x = vec![0.0; n];
    for k in 1..=max_iter {
        let mut diff = 0.0_f64;
        for i in 0..n {
            let mut s = b[i];
            for j in 0..n {
                if j != i { s -= a[i][j] * x[j]; }   // x already holds new entries
            }
            let xi = s / a[i][i];
            diff = diff.max((xi - x[i]).abs());
            x[i] = xi;
        }
        if diff <= tol { return (x, k); }
    }
    (x, max_iter)
}

fn main() {
    let a = vec![vec![4.0, 2.0, -1.0], vec![5.0, -10.0, 2.0], vec![-2.0, 3.0, -7.0]];
    let b = vec![9.0, 8.0, 3.0];
    let (x, it) = gauss_seidel(&a, &b, 1e-10, 200);
    println!("x = {:?}  iterations = {}", x, it);
}
